func_sim/func_memory/func_memory.cpp

```diff
@@ -17,33 +19,80 @@ FuncMemory::FuncMemory( const char* executable_file_name,
                         uint64 page_bits,
                         uint64 offset_bits)
 {
-    // put your code here
+    page_size = 2 << offset_bits;
+    set_size = 2 << page_bits;
+    this->addr_size = addr_size;
+    this->page_bits = page_bits;
+    this->offset_bits = offset_bits;
+    
+    assert( executable_file_name);
+    vector<ElfSection> sections_array;
+    ElfSection::getAllElfSections( executable_file_name, sections_array);
+    assert( sections_array.empty());
+    for ( int i = 0; i < sections_array.size(); i++)
+    {
+        if ( !strcmp( sections_array[i].name, ".text"))
+        {
+            text_start = sections_array[i].start_addr;
+        }
+        
+        uint64 start_addr = sections_array[i].start_addr;
+        
+        for ( uint64 cur_addr = start_addr; cur_addr - start_addr < sections_array[i].size; cur_addr += 1)
+        {
+            Address parsed_addr( cur_addr, addr_size, page_bits, offset_bits);
+            memory[parsed_addr.set]
+                  [parsed_addr.page]
+                  [parsed_addr.offset] = sections_array[i].content[cur_addr - start_addr];
+        }
+    }
 }
 
-FuncMemory::~FuncMemory()
+Address::Address ( uint64 addr,
+                   uint64 addr_size,
+                   uint64 page_bits,
+                   uint64 offset_bits)
 {
-    // put your code here
+    offset = addr & ( ( 2 << offset_bits) - 1);
+    page   = addr & ( ( 2 << ( offset_bits + page_bits)) - 1) - offset;
+    set    = addr & - offset - page;
 }
 
+
 uint64 FuncMemory::startPC() const
 {
-    // put your code here
+    return text_start;
 }
 
-uint64 FuncMemory::read( uint64 addr, unsigned short num_of_bytes) const
+uint64 FuncMemory::read( uint64 addr, unsigned short num_of_bytes)
```

Comment:
For that case C++11 introduces `const mapped_type& at(const key_type& k) const` method: http://www.cplusplus.com/reference/map/map/at/

If you haven't enabled C++11 in your compiler, you'd better use `const_iterator find(const key_type& k) const` method — it's a little bit more tricky to use it (you have to access `std::pair` with `->second` etc.), but object won't be modified.

By the way, `std::vector` does have `const T& operator[](size_t index) const` method. The problem with `std::map` is that `operator[]` *must* return a reference to an object. If there is no object with that key, new object will be created, therefore container will be modified.